walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,121 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+  /**
+   * create a query to select available time slots based on request.
+   * @param events existing events for attendees to participate
+   * @param request request of meetings with specific time and attendees
+   * @return a collection of available time slots
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long requiredDuration = request.getDuration();
+
+    // Check edge cases when the required duration is too long
+    if (requiredDuration > TimeRange.WHOLE_DAY.duration()) {
+      return new ArrayList<>();
+    }
+
+    // Find the available time slots for required attendees and optional attendees
+    ArrayList<TimeRange> requiredTime = new ArrayList<>(
+        Collections.singletonList(
+            TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true)
+        )
+    );
+    ArrayList<TimeRange> optionalTime = new ArrayList<>(
+        Collections.singletonList(
+            TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true)
+        )
+    );
+    for (Event event : events) {
+      TimeRange eventTime = event.getWhen();
+      if (!Collections.disjoint(event.getAttendees(), requiredAttendees)) {
+        requiredTime = splitToTwoRange(eventTime, requiredDuration, requiredTime);
+      }
+      if (!Collections.disjoint(event.getAttendees(), optionalAttendees)) {
+        optionalTime = splitToTwoRange(eventTime, requiredDuration, optionalTime);
+      }
+    }
+
+    // Combine the required and optional time slots to find the intersection
+    ArrayList<TimeRange> result = getTimeIntersection(requiredTime, optionalTime, requiredDuration);
+    return (result.isEmpty() && !requiredAttendees.isEmpty()) ? requiredTime : result;
+  }
+
+  /**
+   * * Split an existing time slot list into two parts to avoid occupied time period.
+   * @param occupied unaccessible time slots
+   * @param duration minimum duration of each time slot
+   * @param remainingTime existing list of available slots
+   * @return an list of modified available time slots without overlapping with occupied period
+   */
+  private ArrayList<TimeRange> splitToTwoRange(
```

Comment:
SplitToTwoRange doesn't make too much sense. How about ReduceAvailability or something like that?